Cudos 1216 extend cudosjs to accommodate the gravity module

BogoLime opened the pull request:
1. Generated .ts types based on the .proto files in the Gravity-bridge repo
2. Implemented User Messages 
  + `SendToEth`, `CancelSendToEth` 
  + also included `MsgSetMinFeeTransferToEth` just in case
3. Implemented all 21 Queries
4. Added additional helper functions for validation in src/utils/checks.ts:
 + `checkValidETHAddress` - validate the address parameters used for ERC20, receiver and contracts addresses
 + `checkValidValidatorAddress`

5. Refactored the `CudosStargateClient class` - which contains all the query methods
 + Combining all query methods from Group module, NFT module and now Gravity module( 21 methods alone ) into this single class in my opinion makes it less structured and more difficult to maintain and read, especially if we keep on adding new modules(each one with ten's of methods)
 + I separated the queries corresponding to the each module in it's own class, which get's imported in the main `CudosStargateClient class`. This changes the Interface of the class ( it's less intuitive ), but makes it more manageable.

 6. Wrote the tests for the messages

BogoLime commented:
Currently, there is a bug with the CancelSendToETH Transaction that is only fixed inside the the cudos-dev branch of the https://github.com/CudoVentures/cudos-node.git. Those fixes have still not been applied to the public and  private test-nets and cudos-blast respectively ( something that is planned to happen soon with the next update). That's why the current tests for gravity will only work on a local node, instantiated (trough  cudos-builders) with the settings from cudos-dev branch. 

NOTE: The test script inside package.json should work as expected for all tests, once the fix is applied to all testing environments we're using.

mlukanova commented on local-accounts.json:
```diff
@@ -0,0 +1,62 @@
+{
```
I guess this is the account file for cudos blast? I don't see a reason to keep it in the root directory of the project. If it is required for the tests move it there.

BogoLime replied:
I've just overlooked that one. It is not needed anywhere in the project. It is now removed.

mlukanova commented on .gitignore:
```diff
@@ -37,4 +37,9 @@ coverage/
+package-lock.json
+
+
+explanation.md
```
Seems like local specific files are added to the .gitignore. Do we need those for every user of the project?

BogoLime replied:
This is a file I created for myself only. It has been removed from the repo and `.gitignore` respectively.

BogoLime commented:
## NEW COMMIT

#### Refactored the code and added some new util functions, that should make some operations easier and the package generally more useful:
 
 1. Exported some additional imports from cosmosJS - mainly related to math and it's types
 2. Created an additional `Registry` ( called `FullRegistry`) - it contains all Msg Proto-types( generated from the `.proto` files) of the custom modules( nft, gravity, group), and adds them to the already existing in cosmJS `DefaultRegistryTypes`(for the standard modules). WHY?
    - Using the queries already implemented for the different modules in `CudosStargateClient`     decodes the responses, BUT executing custom queries( calling default `StargateClient` functions such as `getTx` / `searchTx` which also support filters) returns encoded trx.messages (of the custom modules ), which are  not registered in the defaultRegistry.
3. Added the `FullRegistry` as a property to the `CudosStargateClient`, which now allows it to decode Msg's from the custom modules trough custom queries.
4. Added `decodeQryResponse` function to `CudosStargateClient`, which decodes the responses from the gRPC calls for the custom queries. It takes the response from the query, loops trough the messages, decodes them( using the `FullRegistry`) and returns the decoded query.
5. Added a `generateMsg` factory function - constructing messages is now made easier trough this function. It takes the name of the message as a first parameter and an object with the relevant arguments as a second one. Example: `generateMsg('msgSendToEth', {sender,ethDest,amount})`
    - The added benefit here is that upon typing the function will display the available types of messages and respectively the type of arguments it get's - No need to remember the gRPC endpoints for each message
6. Exported the msg/query proto-types from cudosJS - to be used as types

maptuhec commented on src/stargate/modules/nft/clients/queryClient.ts:
```diff
@@ -0,0 +1,100 @@
+        checkValidNftDenomId(denomId);
+
+        if (owner !== undefined) {
+            checkValidAddress(owner);
```
would't it be better if this check is directly into the `checkValidAddress` function?

BogoLime replied:
Actually, this check is specific to this particular method , and not the `checkValidAddress ` function itself.  Reason is that here  the `owner` parameter is optional and the validity check should be conducted only if such a parameter has been provided. `checkValidAddress` is also used in multiple methods where such a check is not needed. A cleaner way for this method might be to assign a default value to `owner` parameter, but this would change the method API a bit (because of parameter reordering), which I think is unnecessary.

maptuhec commented on src/stargate/modules/nft/clients/queryClient.ts:
```diff
@@ -0,0 +1,100 @@
+        checkValidAddress(owner);
+
+        if (denomId !== undefined) {
+            checkValidNftDenomId(denomId);
```
same here?

BogoLime replied:
Same as above
